File: ruby/snake_game/snake_game.rb
require "gosu"

class SnakeGame < Gosu::Window
  WIDTH, HEIGHT = 400, 400
  TILE = 10

  WIDTH_IN_TILE = WIDTH/TILE

  def initialize
    super WIDTH, HEIGHT, false, 1000/15

    @snake = Snake.new
    @fruit = Fruit.new
  end

  def update
    if @snake.x == @fruit.x && @snake.y == @fruit.y
      @snake.increase
      @fruit.regenerate
    end

    @fruit.update
    @snake.update

    @snake.pos[0..-2].each do |x,y|
      @snake.reset if x == @snake.x && y == @snake.y
    end
  end

  def draw
    @fruit.draw
    @snake.draw
  end

  def button_down(id)
    case id
    when Gosu::KbUp    then @snake.up
    when Gosu::KbDown  then @snake.down
    when Gosu::KbLeft  then @snake.left
    when Gosu::KbRight then @snake.right
    end
  end
end

class Snake
  attr_accessor :x, :y
  attr_reader :pos

  MIN_TAIL = 3

  def initialize
    @x = 10
    @y = 10

    @velx = 0
    @vely = 0

    @tail = MIN_TAIL

    @pos = []
  end

  def update
    @x += @velx
    @y += @vely

    @x = 0 if @x > SnakeGame::WIDTH_IN_TILE
    @x = SnakeGame::WIDTH_IN_TILE if @x < 0
    @y = 0 if @y > SnakeGame::WIDTH_IN_TILE
    @y = SnakeGame::WIDTH_IN_TILE if @y < 0

    @pos << [@x, @y]

    @pos.shift until @pos.size <= @tail
  end

  def draw
    @pos.each do |x,y|
      Gosu.draw_rect(
        x * SnakeGame::TILE,
        y * SnakeGame::TILE,
        SnakeGame::TILE - 1,
        SnakeGame::TILE - 1,
        Gosu::Color::GREEN
      )
    end
  end

  def reset
    @tail = MIN_TAIL
  end

  def increase
    @tail += 1
  end

  def up;    @velx = 0;  @vely = -1; end
  def down;  @velx = 0;  @vely = 1;  end
  def left;  @velx = -1; @vely = 0;  end
  def right; @velx = 1;  @vely = 0;  end
end

class Fruit
  attr_accessor :x, :y

  def initialize
    @x = 15
    @y = 15
  end

  def update
  end

  def draw
    Gosu.draw_rect(
      @x * SnakeGame::TILE,
      @y * SnakeGame::TILE,
      SnakeGame::TILE - 1,
      SnakeGame::TILE - 1,
      Gosu::Color::RED
    )
  end

  def regenerate
    @x = (rand * SnakeGame::TILE).floor
    @y = (rand * SnakeGame::TILE).floor
  end
end

SnakeGame.new.show

File: python/snakegame.py
# pip install pygame # game itsel
# pip install mypy   # type checking

from typing import List, Tuple

import pygame  # type: ignore
from random import randint  # type: ignore

WINDOW_WIDTH = 400
WINDOW_HEIGHT = 400
FPS = 10


class Snake(object):
    WIDTH = 20
    HEIGHT = 20
    MIN_TAIL = 3
    WIDTH_TILES_PER_WINDOW = WINDOW_WIDTH / WIDTH
    HEIGHT_TILES_PER_WINDOW = WINDOW_HEIGHT / HEIGHT

    def __init__(self) -> None:
        self.pos_x = 1
        self.pos_y = 1
        self.vel_x = 0
        self.vel_y = 0

        self.tail = self.MIN_TAIL
        self.trail = []  # type: List[Tuple[int, int]]

    def update(self, events: List[pygame.event.Event]) -> None:
        self.pos_x += self.vel_x
        self.pos_y += self.vel_y

        self.vel_x, self.vel_y = self.handle_mov(events)
        self.check_borders()

        self.trail.append((self.pos_x, self.pos_y))

        while(len(self.trail) > self.tail):
            self.trail.pop(0)

    def draw(self, screen: pygame.Surface) -> None:
        for x, y in self.trail:
            pygame.draw.rect(
                screen,
                pygame.color.Color("red"),
                (
                    x * self.WIDTH,
                    y * self.HEIGHT,
                    self.WIDTH - 1,
                    self.HEIGHT - 1,
                ), 0
            )

    def check_borders(self) -> None:
        if self.pos_x > self.WIDTH_TILES_PER_WINDOW - 1:
            self.pos_x = 0
        if self.pos_x < 0:
            self.pos_x = int(self.WIDTH_TILES_PER_WINDOW - 1)
        if self.pos_y > self.HEIGHT_TILES_PER_WINDOW - 1:
            self.pos_y = 0
        if self.pos_y < 0:
            self.pos_y = int(self.HEIGHT_TILES_PER_WINDOW - 1)

    def handle_mov(self, events: List[pygame.event.Event]) -> Tuple[int, int]:
        movement = (self.vel_x, self.vel_y)

        for ev in events:
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_UP:
                    movement = (0, -1)
                elif ev.key == pygame.K_DOWN:
                    movement = (0, 1)
                elif ev.key == pygame.K_LEFT:
                    movement = (-1, 0)
                elif ev.key == pygame.K_RIGHT:
                    movement = (1, 0)

        return movement

    def increase(self) -> None:
        self.tail += 1

    def reset(self) -> None:
        self.tail = self.MIN_TAIL

    def tail_collision(self) -> bool:
        # we're not considering the last item because it will always
        # match with the current one being checked
        for x, y in self.trail[:-2]:
            if x == self.pos_x and y == self.pos_y:
                return True

        return False


class Fruit(object):
    WIDTH = 20
    HEIGHT = 20

    def __init__(self) -> None:
        self.pos_x = 5
        self.pos_y = 5

    def update(self) -> None:
        pass

    def draw(self, screen: pygame.Surface) -> None:
        pygame.draw.rect(
            screen,
            pygame.color.Color("green"),
            (
                self.pos_x * self.WIDTH,
                self.pos_y * self.HEIGHT,
                self.WIDTH - 1,
                self.HEIGHT - 1,
            )
        )

    def regenerate(self) -> None:
        self.pos_x = randint(0, self.WIDTH - 1)
        self.pos_y = randint(0, self.HEIGHT - 1)


def game_loop(screen: pygame.Surface, font: pygame.font.Font) -> None:
    clock = pygame.time.Clock()
    snake = Snake()
    fruit = Fruit()

    # sound = pygame.mixer.Sound("/tmp/peep.ogg")

    while True:
        events = pygame.event.get()
        handle_quit(events)

        # paint the entire screen
        screen.fill(pygame.color.Color("black"))

        if snake.pos_x == fruit.pos_x and snake.pos_y == fruit.pos_y:
            # sound.play()
            snake.increase()
            fruit.regenerate()

        if snake.tail_collision():
            snake.reset()

        # paint the snake
        snake.update(events)
        snake.draw(screen)

        # paint the fruit
        fruit.update()
        fruit.draw(screen)

        score = font.render(
            "score: %s" % snake.tail, 1, pygame.color.Color("white"))
        screen.blit(score, (0, 0))

        # update the screen
        pygame.display.update()

        # deal with frames per second
        clock.tick(FPS)


def handle_quit(events: List[pygame.event.Event]) -> None:
    for ev in events:
        if ev.type == pygame.QUIT:
            exit(0)
        if ev.type == pygame.KEYDOWN and ev.key == pygame.K_ESCAPE:
            exit(0)


def run() -> None:
    pygame.font.init()
    pygame.init()

    screen = pygame.display.set_mode(
        (WINDOW_WIDTH, WINDOW_HEIGHT),
        pygame.HWSURFACE
    )
    pygame.display.set_caption("Snakegame")
    font = pygame.font.SysFont("monospace", 15)

    game_loop(screen, font)


run()
